src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -85,6 +85,14 @@ public static bool AreTypesAssignable(TypeSymbol sourceType, TypeSymbol targetTy
                     // string literals can be assigned to strings
                     return targetType.Name == LanguageConstants.String.Name;
 
+                case PrimitiveType _ when targetType.Name == LanguageConstants.StringTypeName && sourceType is KeyVaultSecretReferenceType:
+                    // key vault reference type can be assigned only to string type parameters with secure decorator
+                    return targetType.ValidationFlags.HasFlag(TypeSymbolValidationFlags.AllowKeyVaultSecretReferenceAssignment);
+
+                case KeyVaultSecretReferenceType _:
+                    //key vault secret reference type is source type only.
+                    return false;
+
```

Comment:
Couldn't we just use `LanguageConstants.SecureString` here by augmenting string assignment with the requirement that you can't assign secure to insecure? Or does that introduce other problems?

Reply:
It did introduce problem with ambiguous errors like 'cannot assign string to string' when using against non-secure parameter.

It was also being allowed inside string interpolation.